### DISCUSSION TITLE
Latency issues with alexa + matter + esp32c3 + occupancy sensor

### DISCUSSION THREAD

**Discussion by S-A-M-J** [arendst/Tasmota#19232] [Support request] [answered]
As mentioned in the title i have an occupancy sensor of type LD2410 connected to an esp32c3 that runs tasmota. It connects over matter to alexa and shows up as occupancy sensor. So far all of that works.
However the latency between the sensor noticing movement and alexa reacting is about 5-10 seconds. Looking through the code i was unable to find the reason for this rather large delay. 

I have checked the sensor and it works fine and without a delay. From the code [here](https://tasmota.github.io/docs/Matter-Internals/#what-happens-when-an-attribute-is-updated) i gather that the value will be noticed at latest 250ms after change by the matter firmware. Which leaves the communication between the esp32 c3 and the alexa echo dot v3 and the internal communication of the echo dot as the potential sources. 

Maybe someone can shine a light on this problem or has an idea how to fix it?

**Comment by sfromis** [comment]
I'd suggest to activate log level 3, and look for timing details in the logs. That's much more interesting than merely observing that things are slow, and than diving into the code before analyzing available clues.

**Reply by S-A-M-J** [reply]
Thank you for the suggestion. I changed the WebLog level to 3 and observed the following:

1. 21:08:11.891 SRC: Switch
2. 21:08:11.896 RSL: RESULT = {"POWER":"OFF"}
3. 21:08:11.897 RSL: POWER = OFF
4. 21:08:12.430 CFG: Saved, Count 80, Bytes 4096
5. 21:08:16.554 MTR: <Sub_Data  ( 61062) sub=23759
6. 21:08:16.577 MTR: >Read_Attr ( 61062) [01]0406/0000 (Occupancy) - true
7. 21:08:16.674 MTR: .Sub_Done  (      ) sub=23759
8. 21:08:28.798 SRC: Switch
9. 21:08:28.804 RSL: RESULT = {"POWER":"ON"}
10. 21:08:28.806 RSL: POWER = ON
11. 21:08:29.458 CFG: Saved, Count 81, Bytes 4096
12. 21:08:32.028 MTR: <Sub_Data  ( 61062) sub=23759
13. 21:08:32.051 MTR: >Read_Attr ( 61062) [01]0406/0000 (Occupancy) - false
14. 21:08:32.153 MTR: .Sub_Done  (      ) sub=23759

I think this is basically the sensor "seeing" a person in Line 1 and then not "seeing" a person in Line 8. 
However the communication over matter shown in line 5-7 and line 11-14 is done in both occassions roughly 4-5 seconds after the initial reading. 
This i think confirms that a substantial part of the delay is from the esp32-c3 or the tasmota firmware. 
Also interesting is imo that it takes about half a second to save the data from the sensor read which i think is quite long. So far i couldn't make out the specific reasons for this delay in the firmware but i'll keep looking. Maybe someone has a good idea?

P.S.: I love the Tasmota project and want to thank everyone contributing alot...

**Reply by S-A-M-J** [reply]
I did some further analysis and for 22 events (meaning the sensor switched states)
the results are the following:
Average time difference: 2.930 seconds
Maximum time difference: 5.527 seconds
Minimum time difference: 0.723 seconds
Number of entries: 22

**Reply by S-A-M-J** [reply]
Looking at this data:
Switch input recognized time ,sub done time ,Time Difference
20:53:10.405,20:53:13.903,3.498
20:53:22.207,20:53:24.180,1.973
20:53:32.388,20:53:34.609,2.221
20:53:48.686,20:53:49.661,0.975
20:54:15.119,20:54:20.646,5.527
20:54:56.275,20:55:01.184,4.909
20:55:20.921,20:55:21.649,0.728
20:55:36.178,20:55:36.901,0.723
20:56:00.201,20:56:02.649,2.448
20:56:12.421,20:56:13.155,0.734
20:57:08.808,20:57:09.657,0.849
20:57:35.814,20:57:40.397,4.583
20:58:26.914,20:58:31.403,4.489
20:58:39.715,20:58:41.440,1.725
21:08:11.891,21:08:16.674,4.783
21:08:28.798,21:08:32.153,3.355
21:42:50.222,21:42:54.145,3.923
21:42:57.338,21:42:59.430,2.092
21:47:34.307,21:47:39.650,5.343
21:47:59.118,21:48:00.148,1.03
21:48:00.735,21:48:05.420,4.685
21:48:42.546,21:48:46.409,3.863

it seems like the there is a minimum of about 5 seconds between each sub message. This coincides with the polling timeframe declared [here](https://tasmota.github.io/docs/Matter/#getting-started). Although i don't quite understand whether the behaviour mentioned in the docs applies to this scenario. So maybe this is just a coincidence? also in the Tasmota-development\lib\libesp32\berry_matter\src\embedded\Matter_IM_Subscription.be file a min_interval is mentioned which i suspect could be the reason for the 5 sec delay. However i'm not sure where this value is set or what it is initialized to.

**Reply by sfromis** [reply]
Since the docs say "every 5 seconds for sensors", and your device acts as an occupancy sensor, it does seem plausible that the polling interval is what causes the latency up to ~5 seconds. Not having worked with those parts, I cannot say for sure, or if you can set it up differently for fast reaction, like not setting it up as an occupancy sensor.

**Reply by S-A-M-J** [reply]
Yes, it seems that the firmware only checks for new messages to send from sensors every five seconds. Honestly i dont quite understand that decision in terms of an occupancy sensor as this produces an average delay of 2.5 seconds which is just too long in my opinion. 

Hopefully someone can point me to where i can adjust this value or why it is set so high?

Thank you sfromis for helping me this far :)

**Reply by sfromis** [reply]
I'd guess that it is high because it is seen as a *sensor*. Most sensors are for things like temperature, where a few seconds more or less is not an issue, and thus it makes perfect sense that Matter queries at such intervals; in this context, 5 seconds is actually pretty fast. Thus, I'd expect a swift reaction to mean that it should not be configured as a **sensor**, but something else..... As indicated already I have no specific suggestion, but it would appear that Contact or OnOff might be plausible things to try. Still, I'm a bit surprised that an *occupancy* sensor is not polled faster, but ideally I'd say that wanting swift reactions would mean that it should just not be *polled*, but instead having the Matter endpoint device initiate a transmission.

To be clear, what I write here is basically just speculation/guessing without knowing how to set things up.

**Comment by s-hadinger** [comment]
This is an arbitrary value that I thought made sense for sensors like Temperature.

The parameter is here: https://github.com/arendst/Tasmota/blob/222b47627f9cce81733ed10c79d8beb4516266ef/lib/libesp32/berry_matter/src/embedded/Matter_Plugin_Sensor_Occupancy.be#L35

For a contact, I agree that polling can be much more aggressive without harm. Would every second be ok? Or even shorter?

**Reply by S-A-M-J** [reply]
Ah that makes sense. I think 500ms would be a good value if the esp can handle that? Is there a specific reason you went with polling instead of sending a message on Event basis?

**Reply by S-A-M-J** [reply]
I think 500ms is good since there seems to be an unrelated to the update_time value delay of about 500ms according to the analysis I made. That would make the delay on average 750ms which I think is ok.

**Reply by s-hadinger** [reply]
`Status 8` takes 56ms to complete, I would be more comfortable to have 750ms recurrence to avoid any starvation of CPU

**Reply by S-A-M-J** [reply]
Oh ok. Yeah seems reasonable.

**Reply by s-hadinger** [reply]
Using `Status 8` was the easy implementation and shared with Remote Devices. I could be possible to optimize by directly reading Switch values without building the full JSON.

Stay tuned

**Reply by S-A-M-J** [reply]
Very much looking forward to it. Thank you for your for your help btw. 
FYI: I have changed the parameter you suggested to 750 in a custom build and tried it and still get the same average delay. Is it perhaps overwritten at some point?

**Reply by S-A-M-J** [reply]
Apparently my custom build didn't work due to one having to solidify the berry scripts into header files. According to the bottom of [this](https://tasmota.github.io/docs/Matter-Internals/#extending-matter) page this can be done by executing the solidify_all.be script. However here i got stuck as the script throws this error: 

Parsing:  matter.be
Parsing:  Matter_Base38.be
Parsing:  Matter_Commissioning.be
internal_error: Unsupported access to non-builtin global
stack traceback:
        <native>: in native function
        <unknown source>: in function `parse_file`
        <unknown source>: in function `main`

Maybe the script is currently not working in the development branch? I'm unable to figure out exactly what this error means though unfortunately. Maybe someone has an idea?

**Reply by s-hadinger** [reply]
The script is working. It is also working in CI to check that solidification was done right.

**Comment by S-A-M-J** [comment, marked as answer]
Fixed as of now. Thank you @s-hadinger. Works great now.
